Panic on sending results over gRPC to normalizer

Have to do with gRPC message size, most probably a twin on bblfsh/server#58

```
INFO[07-13|15:18:03] Done. All files parsed                   repo=015d36fc-8de9-8fab-e5a9-0c11bdf41e2b success=31 fail=0 skipped=13
INFO[07-13|15:18:03] Done. All files in all repositories parsed repositories=18 files=3722

panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x8 pc=0x16f3737]

goroutine 51 [running]:
github.com/src-d/berserker/extractor.(*RepositoryData).ProtoSize(0x0, 0x2)
	/Users/alex/go/src/github.com/src-d/berserker/extractor/generated.pb.go:418 +0x37
github.com/src-d/berserker/extractor.(*Service_GetRepositoriesDataResponse).ProtoSize(0xc4209ec280, 0x177e160)
	/Users/alex/go/src/github.com/src-d/berserker/extractor/generated.pb.go:464 +0x5b
github.com/src-d/berserker/extractor.(*Service_GetRepositoriesDataResponse).Marshal(0xc4209ec280, 0x219fa80, 0xc4209ec280, 0x3588f58, 0xc4209ec280, 0x1)
	/Users/alex/go/src/github.com/src-d/berserker/extractor/generated.pb.go:337 +0x2f
github.com/src-d/berserker/extractor/vendor/github.com/golang/protobuf/proto.(*Buffer).Marshal(0xc4209f4758, 0x219fa80, 0xc4209ec280, 0xc43019dea0, 0x0)
	/Users/alex/go/src/github.com/src-d/berserker/extractor/vendor/github.com/golang/protobuf/proto/encode.go:264 +0x7a
github.com/src-d/berserker/extractor/vendor/google.golang.org/grpc.protoCodec.marshal(0x1846c60, 0xc4209ec280, 0xc4209f4750, 0xc42d3f57f8, 0x10862c2, 0xed0f96a0b, 0x20f0983e, 0x21ee0e0)
	/Users/alex/go/src/github.com/src-d/berserker/extractor/vendor/google.golang.org/grpc/codec.go:78 +0xe8
github.com/src-d/berserker/extractor/vendor/google.golang.org/grpc.protoCodec.Marshal(0x1846c60, 0xc4209ec280, 0xd, 0x283f5484, 0x21ee0e0, 0xc42d3f5868, 0x106e954)
	/Users/alex/go/src/github.com/src-d/berserker/extractor/vendor/google.golang.org/grpc/codec.go:88 +0x73
github.com/src-d/berserker/extractor/vendor/google.golang.org/grpc.(*protoCodec).Marshal(0x220a7a0, 0x1846c60, 0xc4209ec280, 0x13bd860, 0xc420a50054, 0xed0f96a0b, 0x20f0983e, 0x21ee0e0)
	<autogenerated>:35 +0x59
github.com/src-d/berserker/extractor/vendor/google.golang.org/grpc.encode(0x21a1600, 0x220a7a0, 0x1846c60, 0xc4209ec280, 0x0, 0x0, 0x0, 0x0, 0xed0f96a0b, 0x100000020f0983e, ...)
	/Users/alex/go/src/github.com/src-d/berserker/extractor/vendor/google.golang.org/grpc/rpc_util.go:316 +0x2f9
github.com/src-d/berserker/extractor/vendor/google.golang.org/grpc.(*Server).sendResponse(0xc4209f8000, 0x21a5260, 0xc420a42000, 0xc420a4e000, 0x1846c60, 0xc4209ec280, 0x0, 0x0, 0xc43019de90, 0x0, ...)
	/Users/alex/go/src/github.com/src-d/berserker/extractor/vendor/google.golang.org/grpc/server.go:665 +0xb5
github.com/src-d/berserker/extractor/vendor/google.golang.org/grpc.(*Server).processUnaryRPC(0xc4209f8000, 0x21a5260, 0xc420a42000, 0xc420a4e000, 0xc4209e82d0, 0x21dcc60, 0xc4209dc4e0, 0x0, 0x0)
	/Users/alex/go/src/github.com/src-d/berserker/extractor/vendor/google.golang.org/grpc/server.go:821 +0x12c9
github.com/src-d/berserker/extractor/vendor/google.golang.org/grpc.(*Server).handleStream(0xc4209f8000, 0x21a5260, 0xc420a42000, 0xc420a4e000, 0xc4209dc4e0)
	/Users/alex/go/src/github.com/src-d/berserker/extractor/vendor/google.golang.org/grpc/server.go:997 +0x15a6
github.com/src-d/berserker/extractor/vendor/google.golang.org/grpc.(*Server).serveStreams.func1.1(0xc420a0c090, 0xc4209f8000, 0x21a5260, 0xc420a42000, 0xc420a4e000)
	/Users/alex/go/src/github.com/src-d/berserker/extractor/vendor/google.golang.org/grpc/server.go:561 +0xa9
created by github.com/src-d/berserker/extractor/vendor/google.golang.org/grpc.(*Server).serveStreams.func1
	/Users/alex/go/src/github.com/src-d/berserker/extractor/vendor/google.golang.org/grpc/server.go:562 +0xa1
```

--- a/.gitignore
+++ b/.gitignore
@@ -6,3 +6,4 @@ extractor/vendor
 #*
 .DS_Store
 *.iml
+*~

--- a/extractor/cli/main.go
+++ b/extractor/cli/main.go
@@ -12,7 +12,7 @@ import (
 
 var cpuprofile = flag.String("cpuprofile", "", "write cpu profile to file")
 var memprofile = flag.String("memprofile", "", "write memory profile to this file")
-var n = flag.Uint64("n", 0, "number of repositories, 0 = All from DB")
+var limit = flag.Uint64("limit", 0, "number of repositories, 0 = All from DB")
 
 func main() {
 	flag.Parse()
@@ -25,8 +25,8 @@ func main() {
 		defer pprof.StopCPUProfile()
 	}
 
-	extractorService := extractor.NewService()
-	repos, err := extractorService.GetRerpoData(*n)
+	extractorService := extractor.NewService(*limit)
+	repos, err := extractorService.GetRepositoriesData()
 	checkIfError(err)
 	fmt.Printf("Repos returned: %d\n", len(repos))
 

--- a/extractor/cli/main.go
+++ b/extractor/cli/main.go
@@ -12,6 +12,7 @@ import (
 
 var cpuprofile = flag.String("cpuprofile", "", "write cpu profile to file")
 var memprofile = flag.String("memprofile", "", "write memory profile to this file")
+var n = flag.Uint64("n", 0, "number of repositories, 0 = All from DB")
 
 func main() {
 	flag.Parse()
@@ -25,7 +26,7 @@ func main() {
 	}
 
 	extractorService := extractor.NewService()
-	repos, err := extractorService.GetRepositoriesData()
+	repos, err := extractorService.GetRerpoData(*n)
 	checkIfError(err)
 	fmt.Printf("Repos returned: %d\n", len(repos))
 

--- a/extractor/glide.lock
+++ b/extractor/glide.lock
@@ -1,19 +1,28 @@
-hash: 9102d7bfae648f8494798a476d3853bc6e6cdbc4476bb41462a23ddb85f84d5a
-updated: 2017-07-12T09:30:04.99725705+02:00
+hash: 28df3ad771be2c058e02744a2bfdd93614f928b50a43e2f430590f417b5d111b
+updated: 2017-07-17T10:29:44.265205115+02:00
 imports:
 - name: github.com/bblfsh/sdk
-  version: 67d54901d8ea4929ec5ff4a812825fd6ca6303b0
+  version: c829390f1f87d418285ebacb63bb0b201c6ea903
   subpackages:
   - protocol
   - uast
+- name: github.com/coreos/etcd
+  version: 9ce7bb6a1c15bcffdd011a571dc686e71aabc314
+  subpackages:
+  - auth/authpb
+  - clientv3
+  - clientv3/concurrency
+  - etcdserver/api/v3rpc/rpctypes
+  - etcdserver/etcdserverpb
+  - mvcc/mvccpb
 - name: github.com/gogo/protobuf
-  version: 9df9efe4c742f1a2bfdedf1c3b6902fc6e814c6b
+  version: 100ba4e885062801d56799d78530b73b178a78f3
   subpackages:
   - gogoproto
   - proto
   - protoc-gen-gogo/descriptor
 - name: github.com/golang/protobuf
-  version: 6e4cc92cc905d5f4a73041c1b8228ea08f4c6147
+  version: 0a4f71a498b7c4812f64969510bcb4eca251e33a
   subpackages:
   - proto
   - ptypes/any
@@ -32,13 +41,15 @@ imports:
 - name: github.com/mattn/go-colorable
   version: 3fa8c76f9daed4067e4a806fb7e4dc86455c6d6a
 - name: github.com/mattn/go-isatty
-  version: 57fdcb988a5c543893cc61bce354a6e24ab70022
+  version: fc9e8d8ef48496124e79ae0df75490096eccf6fe
 - name: github.com/mcuadros/go-defaults
   version: e1c978be3307be96ef03a0bd25d719ab50d277e4
 - name: github.com/oklog/ulid
   version: 66bb6560562feca7045b23db1ae85b01260f87c5
+- name: github.com/pkg/errors
+  version: c605e284fe17294bda444b34710735b29d1a9d90
 - name: github.com/satori/go.uuid
-  version: 879c5887cd475cd7864858769793b2ceb0d44feb
+  version: 5bf94b69c6b68ee1b541973bb8e1144db23a194b
 - name: github.com/serenize/snaker
   version: 1c7f65329f6524115af8ca01ab604c00cddf5558
 - name: github.com/sergi/go-diff
@@ -60,7 +71,7 @@ imports:
 - name: github.com/xanzy/ssh-agent
   version: ba9c9e33906f58169366275e3450db66139a31a9
 - name: golang.org/x/crypto
-  version: 3627ff35f31987174dbee61d9d1dcc1c643e7174
+  version: 7f7c0c2d75ebb4e32a21396ce36e87b6dadc91c9
   subpackages:
   - curve25519
   - ed25519
@@ -69,7 +80,7 @@ imports:
   - ssh/agent
   - ssh/knownhosts
 - name: golang.org/x/net
-  version: 3da985ce5951d99de868be4385f21ea6c2b22f24
+  version: b3756b4b77d7b13260a0a2ec658753cf48922eac
   subpackages:
   - context
   - http2
@@ -79,11 +90,11 @@ imports:
   - lex/httplex
   - trace
 - name: golang.org/x/sys
-  version: 8f0908ab3b2457e2e15403d3697c9ef5cb4b57a9
+  version: 4cd6d1a821c7175768725b55ca82f14683a29ea4
   subpackages:
   - unix
 - name: golang.org/x/text
-  version: 4ee4af566555f5fbe026368b75596286a312663a
+  version: 836efe42bb4aa16aaa17b9c155d8813d336ed720
   subpackages:
   - secure/bidirule
   - transform
@@ -101,11 +112,11 @@ imports:
   - internal/modules
   - internal/remote_api
 - name: google.golang.org/genproto
-  version: d80a6e20e776b0b17a324d0ba1ab50a39c8e8944
+  version: b0a3dcfcd1a9bd48e63634bd8802960804cf8315
   subpackages:
   - googleapis/rpc/status
 - name: google.golang.org/grpc
-  version: 79f73d62e5082ee8d6b86a2487fd6267c1495fcd
+  version: b15215fb911b24a5d61d57feec4233d610530464
   subpackages:
   - codes
   - credentials
@@ -125,13 +136,8 @@ imports:
   subpackages:
   - stack
   - term
-- name: gopkg.in/mgo.v2
-  version: 3f83fa5005286a7fe593b055f0d7771a7dce4655
-  subpackages:
-  - bson
-  - internal/json
 - name: gopkg.in/src-d/core-retrieval.v0
-  version: bf9b3d791c9df25da40efe0950c647713cc13f50
+  version: 83bf9ada96f00497f460c36879d53c980f060e18
   subpackages:
   - model
   - repository
@@ -140,14 +146,16 @@ imports:
   subpackages:
   - model
 - name: gopkg.in/src-d/enry.v1
-  version: ad46ae50f84afb299328225ecd82c9d3f6acb9e8
+  version: b2b40bbfc59fa9ab248d8662edca78839195a7a6
   subpackages:
+  - data
   - internal/tokenizer
 - name: gopkg.in/src-d/framework.v0
-  version: 06c61f8f6a8913f3d4d465bab758ff06c0b3d920
+  version: b7e5e412473a20a56c0543f898624f024b164087
   subpackages:
   - configurable
   - database
+  - lock
   - queue
 - name: gopkg.in/src-d/go-billy-siva.v3
   version: b15b35bba22ab049715a74125fb100faee83cd00
@@ -159,9 +167,10 @@ imports:
   - helper/polyfill
   - osfs
   - util
+- name: gopkg.in/src-d/go-errors.v0
+  version: 6f07baca276330431b44ece8ee84ac98167bc4ea
 - name: gopkg.in/src-d/go-git.v4
-  version: ce6f5b7c82fc6c2c4d41880ed6b26f921dd9c1c3
-  repo: https://github.com/src-d/go-git.git
+  version: d3c7400c39f86a4c59340c7a9cda8497186e00fc
   subpackages:
   - config
   - internal/revision
@@ -203,7 +212,7 @@ imports:
   - utils/merkletrie/internal/frame
   - utils/merkletrie/noder
 - name: gopkg.in/src-d/go-kallax.v1
-  version: 2cc6c43422cceee02f3b75a702973fe3834f831f
+  version: 2acd313ffa7ae1f67edec6556600ce2358be3c79
   subpackages:
   - types
 - name: gopkg.in/src-d/go-siva.v1

--- a/extractor/glide.yaml
+++ b/extractor/glide.yaml
@@ -1,30 +1,34 @@
 package: github.com/src-d/berserker/extractor
 import:
+- package: github.com/bblfsh/sdk
+  version: c829390f1f87d418285ebacb63bb0b201c6ea903
+  subpackages:
+  - protocol
 - package: github.com/gogo/protobuf
+  version: ^0.4.0
   subpackages:
   - gogoproto
 - package: github.com/golang/protobuf
   subpackages:
   - proto
-- package: google.golang.org/grpc
-- package: gopkg.in/src-d/enry.v1
-- package: gopkg.in/src-d/go-git.v4
-  repo: https://github.com/src-d/go-git.git
-  version: ce6f5b7c82fc6c2c4d41880ed6b26f921dd9c1c3
+- package: github.com/inconshreveable/log15
+  version: ^2.11.0
+- package: golang.org/x/net
   subpackages:
-  - config
-  - plumbing
-  - plumbing/object
-  - plumbing/storer
-  - plumbing/transport
-  - plumbing/transport/client
-  - plumbing/transport/server
-  - storage/filesystem
-  - utils/ioutil
+  - context
+- package: google.golang.org/grpc
+  version: ^1.4.2
 - package: gopkg.in/src-d/core-retrieval.v0
-  version: bf9b3d791c9df25da40efe0950c647713cc13f50
+- package: gopkg.in/src-d/core.v0
   subpackages:
   - model
-  - repository
-- package: github.com/bblfsh/sdk
-  version: 67d54901d8ea4929ec5ff4a812825fd6ca6303b0
+- package: gopkg.in/src-d/enry.v1
+  version: ^1.3.0
+- package: gopkg.in/src-d/go-git.v4
+  version: ^4.0.0-rc12
+  subpackages:
+  - plumbing
+  - plumbing/object
+  - storage
+- package: golang.org/x/crypto
+  version: 7f7c0c2d75ebb4e32a21396ce36e87b6dadc91c9

--- a/extractor/server.proteus.go
+++ b/extractor/server.proteus.go
@@ -8,8 +8,8 @@ type extractorServiceServer struct {
 	Service *Service
 }
 
-func NewExtractorServiceServer() *extractorServiceServer {
-	return &extractorServiceServer{NewService()}
+func NewExtractorServiceServer(limit uint64) *extractorServiceServer {
+	return &extractorServiceServer{NewService(limit)}
 }
 func (s *extractorServiceServer) Service_GetRepositoriesData(ctx context.Context, in *Service_GetRepositoriesDataRequest) (result *Service_GetRepositoriesDataResponse, err error) {
 	result = new(Service_GetRepositoriesDataResponse)

--- a/extractor/server/server.go
+++ b/extractor/server/server.go
@@ -28,7 +28,7 @@ func main() {
 		panic(err)
 	}
 
-	grpcServer := grpc.NewServer()
+	grpcServer := grpc.NewServer(grpc.MaxSendMsgSize(extractor.GrpcMaxMsgSize))
 
 	extractor.RegisterExtractorServiceServer(grpcServer, extractor.NewExtractorServiceServer(*limit))
 	log.Info("server started", "address", grpcAddr)

--- a/extractor/server/server.go
+++ b/extractor/server/server.go
@@ -13,11 +13,12 @@ import (
 )
 
 var profiler = flag.Bool("profiler", false, "start CPU & memeory profiler")
+var limit = flag.Uint64("limit", 0, "number of repositories, 0 = All from DB")
 
 func main() {
 	flag.Parse()
 	// TODO parametrize
-	profilerAddr := "localhost:6062"
+	profilerAddr := "localhost:6063"
 	grpcAddr := "localhost:8888"
 
 	startHTTPProfilingMaybe(profilerAddr)
@@ -29,7 +30,7 @@ func main() {
 
 	grpcServer := grpc.NewServer()
 
-	extractor.RegisterExtractorServiceServer(grpcServer, extractor.NewExtractorServiceServer())
+	extractor.RegisterExtractorServiceServer(grpcServer, extractor.NewExtractorServiceServer(*limit))
 	log.Info("server started", "address", grpcAddr)
 
 	err = grpcServer.Serve(lis)

--- a/extractor/service.go
+++ b/extractor/service.go
@@ -24,9 +24,10 @@ import (
 
 type Service struct {
 	bblfshClient protocol.ProtocolServiceClient
+	limit        uint64
 }
 
-func NewService() *Service {
+func NewService(n uint64) *Service {
 	//TODO(bzz): parametrize
 	bblfshAddr := "0.0.0.0:9432"
 	log.Info("Connecting to Bblfsh server", "address", bblfshAddr)
@@ -34,7 +35,8 @@ func NewService() *Service {
 	client := protocol.NewProtocolServiceClient(bblfshConn)
 	checkIfError(err)
 
-	return &Service{bblfshClient: client}
+	log.Info("Limiting number of repositories to N", "N", n)
+	return &Service{bblfshClient: client, limit: n}
 }
 
 //proteus:generate
@@ -45,10 +47,10 @@ func (s *Service) GetRepositoryData(r *Request) (*RepositoryData, error) {
 
 //proteus:generate
 func (s *Service) GetRepositoriesData() ([]*RepositoryData, error) {
-	return s.GetRerpoData(0) //all
+	return s.getRerpoData(s.limit) //all
 }
 
-func (s *Service) GetRerpoData(n uint64) ([]*RepositoryData, error) {
+func (s *Service) getRerpoData(n uint64) ([]*RepositoryData, error) {
 	if n <= 0 {
 		k, err := core.ModelRepositoryStore().Count(model.NewRepositoryQuery().FindByStatus(model.Fetched))
 		if err != nil {
@@ -57,7 +59,7 @@ func (s *Service) GetRerpoData(n uint64) ([]*RepositoryData, error) {
 		}
 		n = uint64(k)
 	}
-	log.Info("Iterating over repositories in DB", "status:fetched", n)
+	log.Info("Iterating over N repositories in DB", "N", n)
 
 	const master = "refs/heads/master"
 	result := make([]*RepositoryData, n)

--- a/extractor/service.go
+++ b/extractor/service.go
@@ -63,7 +63,7 @@ func (s *Service) getRerposData(n uint64) ([]*RepositoryData, error) {
 	log.Info("Iterating over N repositories in DB", "N", n)
 
 	const master = "refs/heads/master"
-	result := make([]*RepositoryData, n)
+	var result []*RepositoryData
 
 	reposNum := 0
 	totalFiles := 0

--- a/extractor/service.go
+++ b/extractor/service.go
@@ -69,7 +69,7 @@ func (s *Service) getRerposData(n uint64) ([]*RepositoryData, error) {
 	totalFiles := 0
 	for masterRefInit, repoMetadata := range findAllFetchedReposWithRef(master, n) {
 		repo, processedFiles, err := s.processRepository(repoMetadata, master, masterRefInit)
-		if err != nil && repo == nil { //partially processed repos are OK
+		if err != nil && repo == nil { // partially processed repos are OK
 			//TODO(bzz): move loggin/error handing here instead of s.processRepository()
 			continue
 		}
@@ -80,6 +80,12 @@ func (s *Service) getRerposData(n uint64) ([]*RepositoryData, error) {
 	}
 
 	log.Info("Done. All files in all repositories parsed", "repositories", reposNum, "files", totalFiles)
+
+	log.Debug("Serializing files in", "repositories", len(result))
+	for _, r := range result {
+		log.Debug("Repository", "ID", r.RepositoryID, "URL", r.URL, "number of files", len(r.Files))
+	}
+
 	return result, nil
 }
 

--- a/extractor/service.go
+++ b/extractor/service.go
@@ -210,10 +210,10 @@ func parseToUast(client protocol.ProtocolServiceClient, fName string, fLang stri
 	//TODO(bzz): take care of non-UTF8 things, before sending them
 	//  - either encode in utf8
 	//  - or convert to base64() and set encoding param
-	req := &protocol.ParseUASTRequest{
+	req := &protocol.ParseRequest{
 		Content:  fContent,
 		Language: fLang}
-	resp, err := client.ParseUAST(context.TODO(), req)
+	resp, err := client.Parse(context.TODO(), req)
 	if err != nil {
 		log.Error("ParseUAST failed on gRPC level", "file", fName, "err", err)
 		return nil, err

--- a/extractor/service.go
+++ b/extractor/service.go
@@ -45,19 +45,26 @@ func (s *Service) GetRepositoryData(r *Request) (*RepositoryData, error) {
 
 //proteus:generate
 func (s *Service) GetRepositoriesData() ([]*RepositoryData, error) {
-	n, err := core.ModelRepositoryStore().Count(model.NewRepositoryQuery().FindByStatus(model.Fetched))
-	if err != nil {
-		log.Error("Could not connect to DB to get the number of 'fetched' repositories", "err", err)
-	} else {
-		log.Info("Iterating over repositories in DB", "status:fetched", n)
+	return s.GetRerpoData(0) //all
+}
+
+func (s *Service) GetRerpoData(n uint64) ([]*RepositoryData, error) {
+	if n <= 0 {
+		k, err := core.ModelRepositoryStore().Count(model.NewRepositoryQuery().FindByStatus(model.Fetched))
+		if err != nil {
+			log.Error("Could not connect to DB to get the number of 'fetched' repositories", "err", err)
+			return nil, err
+		}
+		n = uint64(k)
 	}
+	log.Info("Iterating over repositories in DB", "status:fetched", n)
 
 	const master = "refs/heads/master"
 	result := make([]*RepositoryData, n)
 
 	reposNum := 0
 	totalFiles := 0
-	for masterRefInit, repoID := range findAllFetchedReposWithRef(master) {
+	for masterRefInit, repoID := range findAllFetchedReposWithRef(master, n) {
 		log.Info("Processing repository", "id", repoID)
 		repo := &RepositoryData{
 			RepositoryID: repoID,
@@ -205,9 +212,9 @@ func parseToUast(client protocol.ProtocolServiceClient, fName string, fLang stri
 }
 
 // Collects all Repository metadata in-memory
-func findAllFetchedReposWithRef(refText string) map[model.SHA1]string {
+func findAllFetchedReposWithRef(refText string, n uint64) map[model.SHA1]string {
 	repoStorage := core.ModelRepositoryStore()
-	q := model.NewRepositoryQuery().FindByStatus(model.Fetched)
+	q := model.NewRepositoryQuery().FindByStatus(model.Fetched).Limit(n)
 	rs, err := repoStorage.Find(q)
 	if err != nil {
 		log.Error("Failed to query DB", "err", err)
@@ -215,7 +222,8 @@ func findAllFetchedReposWithRef(refText string) map[model.SHA1]string {
 	}
 
 	repos := make(map[model.SHA1]string)
-	for rs.Next() { // for each Repository
+	for rs.Next() {
+		// for each Repository
 		repo, err := rs.Get()
 		if err != nil {
 			log.Error("Failed to get next row from DB", "err", err)
@@ -229,7 +237,8 @@ func findAllFetchedReposWithRef(refText string) map[model.SHA1]string {
 				break
 			}
 		}
-		if masterRef == nil { // skipping repos \wo it
+		if masterRef == nil {
+			// skipping repos \wo it
 			log.Warn("No reference found ", "repo", repo.ID, "reference", refText)
 			continue
 		}

--- a/extractor/service.go
+++ b/extractor/service.go
@@ -47,7 +47,7 @@ func (s *Service) GetRepositoryData(r *Request) (*RepositoryData, error) {
 
 //proteus:generate
 func (s *Service) GetRepositoriesData() ([]*RepositoryData, error) {
-	return s.getRerpoData(s.limit) //all
+	return s.getRerpoData(s.limit)
 }
 
 func (s *Service) getRerpoData(n uint64) ([]*RepositoryData, error) {
@@ -66,97 +66,107 @@ func (s *Service) getRerpoData(n uint64) ([]*RepositoryData, error) {
 
 	reposNum := 0
 	totalFiles := 0
-	for masterRefInit, repoID := range findAllFetchedReposWithRef(master, n) {
-		log.Info("Processing repository", "id", repoID)
-		repo := &RepositoryData{
-			RepositoryID: repoID,
-			URL:          "", //TODO(bzz): add repo url!
-			Files:        make([]File, 100),
-		}
-
-		rootedTransactioner := core_retrieval.RootedTransactioner()
-		tx, err := rootedTransactioner.Begin(plumbing.Hash(masterRefInit))
+	for masterRefInit, repoMetadata := range findAllFetchedReposWithRef(master, n) {
+		repo, processedFiles, err := s.processRepository(repoMetadata, master, masterRefInit)
 		if err != nil {
-			log.Error("Failed to begin tx for rooted repo", "id", repoID, "hash", masterRefInit, "err", err)
 			continue
 		}
 
-		tree, err := gitOpenGetTree(tx.Storer(), repoID, masterRefInit, master)
-		if err != nil {
-			log.Error("Failed to open&get tree from rooted repo", "id", repoID, "hash", masterRefInit, "err", err)
-			_ = tx.Rollback()
-			continue
-		}
-
-		skpFiles := 0
-		sucFiles := 0
-		errFiles := 0
-		err = tree.Files().ForEach(func(f *object.File) error {
-			i := (skpFiles + sucFiles + errFiles) % 1000
-			batch := (skpFiles + sucFiles + errFiles) / 1000
-			if i == 0 && batch != 0 {
-				fmt.Printf("\t%d000 files...\n", batch)
-			}
-
-			// discard vendoring with enry
-			if enry.IsVendor(f.Name) || enry.IsDotFile(f.Name) ||
-				enry.IsDocumentation(f.Name) || enry.IsConfiguration(f.Name) {
-				skpFiles++
-				return nil
-			} //TODO(bzz): filter binaries like .apk and .jar
-
-			// detect language with enry
-			fContent, err := f.Contents()
-			if err != nil {
-				log.Warn("Failed to read", "file", f.Name, "err", err)
-				errFiles++
-				return nil
-			}
-
-			fLang := enry.GetLanguage(f.Name, []byte(fContent))
-			if err != nil {
-				log.Warn("Failed to detect language", "file", f.Name, "err", err)
-				errFiles++
-				return nil
-			}
-			//log.Debug(fmt.Sprintf("\t%-9s blob %s    %s", fLang, f.Hash, f.Name))
-
-			// Babelfish -> UAST (Python, Java)
-			if strings.EqualFold(fLang, "java") || strings.EqualFold(fLang, "python") {
-				uast, err := parseToUast(s.bblfshClient, f.Name, strings.ToLower(fLang), fContent)
-				if err != nil {
-					errFiles++
-					return nil
-				}
-
-				sucFiles++
-				file := File{
-					Language: fLang,
-					Path:     f.Name,
-					UAST:     *uast,
-				}
-				repo.Files = append(repo.Files, file)
-			}
-			return nil
-		})
-		if err != nil {
-			log.Error("Failed to iterate files in", "repo", repoID)
-			continue
-		}
 		result = append(result, repo)
-
-		log.Info("Done. All files parsed", "repo", repoID, "success", sucFiles, "fail", errFiles, "skipped", skpFiles)
 		reposNum++
-		totalFiles = totalFiles + sucFiles + errFiles + skpFiles
-
-		err = tx.Rollback()
-		if err != nil {
-			log.Error("Failed to rollback tx for rooted repo", "repo", repoID, "err", err)
-			continue
-		}
+		totalFiles = totalFiles + processedFiles
 	}
+
 	log.Info("Done. All files in all repositories parsed", "repositories", reposNum, "files", totalFiles)
 	return result, nil
+}
+
+func (s *Service) processRepository(repoMetadata *model.Repository, master string, masterRefInit model.SHA1) (*RepositoryData, int, error) {
+	repoID := repoMetadata.ID.String()
+	log.Info("Processing repository", "id", repoID)
+	repo := &RepositoryData{
+		RepositoryID: repoID,
+		URL:          repoMetadata.Endpoints[0], //no endpoints?
+		Files:        make([]File, 100),
+	}
+
+	tx, err := core_retrieval.RootedTransactioner().Begin(plumbing.Hash(masterRefInit))
+	if err != nil {
+		log.Error("Failed to begin tx for rooted repo", "id", repoID, "hash", masterRefInit, "err", err)
+		return nil, 0, err
+	}
+
+	tree, err := gitOpenGetTree(tx.Storer(), repoID, masterRefInit, master)
+	if err != nil {
+		log.Error("Failed to open&get tree from rooted repo", "id", repoID, "hash", masterRefInit, "err", err)
+		_ = tx.Rollback()
+		return nil, 0, err
+	}
+
+	skpFiles := 0
+	sucFiles := 0
+	errFiles := 0
+	err = tree.Files().ForEach(func(f *object.File) error {
+		i := (skpFiles + sucFiles + errFiles) % 1000
+		batch := (skpFiles + sucFiles + errFiles) / 1000
+		if i == 0 && batch != 0 {
+			fmt.Printf("\t%d000 files...\n", batch)
+		}
+
+		// discard vendoring with enry
+		if enry.IsVendor(f.Name) || enry.IsDotFile(f.Name) ||
+			enry.IsDocumentation(f.Name) || enry.IsConfiguration(f.Name) {
+			skpFiles++
+			return nil
+		} //TODO(bzz): filter binaries like .apk and .jar
+
+		// detect language with enry
+		fContent, err := f.Contents()
+		if err != nil {
+			log.Warn("Failed to read", "file", f.Name, "err", err)
+			errFiles++
+			return nil
+		}
+
+		fLang := enry.GetLanguage(f.Name, []byte(fContent))
+		if err != nil {
+			log.Warn("Failed to detect language", "file", f.Name, "err", err)
+			errFiles++
+			return nil
+		}
+		//log.Debug(fmt.Sprintf("\t%-9s blob %s    %s", fLang, f.Hash, f.Name))
+
+		// Babelfish -> UAST (Python, Java)
+		if strings.EqualFold(fLang, "java") || strings.EqualFold(fLang, "python") {
+			uast, err := parseToUast(s.bblfshClient, f.Name, strings.ToLower(fLang), fContent)
+			if err != nil {
+				errFiles++
+				return nil
+			}
+
+			sucFiles++
+			file := File{
+				Language: fLang,
+				Path:     f.Name,
+				UAST:     *uast,
+			}
+			repo.Files = append(repo.Files, file)
+		}
+		return nil
+	})
+	if err != nil {
+		log.Error("Failed to iterate files in", "repo", repoID)
+		return nil, sucFiles + errFiles + skpFiles, err
+	}
+
+	log.Info("Done. All files parsed", "repo", repoID, "success", sucFiles, "fail", errFiles, "skipped", skpFiles)
+	err = tx.Rollback()
+	if err != nil {
+		log.Error("Failed to rollback tx for rooted repo", "repo", repoID, "err", err)
+		return nil, sucFiles + errFiles + skpFiles, err
+	}
+
+	return repo, sucFiles + errFiles + skpFiles, nil
 }
 
 func gitOpenGetTree(txStorer storage.Storer, repoID string, masterRefInit model.SHA1, master string) (*object.Tree, error) {
@@ -214,7 +224,7 @@ func parseToUast(client protocol.ProtocolServiceClient, fName string, fLang stri
 }
 
 // Collects all Repository metadata in-memory
-func findAllFetchedReposWithRef(refText string, n uint64) map[model.SHA1]string {
+func findAllFetchedReposWithRef(refText string, n uint64) map[model.SHA1]*model.Repository {
 	repoStorage := core.ModelRepositoryStore()
 	q := model.NewRepositoryQuery().FindByStatus(model.Fetched).Limit(n)
 	rs, err := repoStorage.Find(q)
@@ -223,7 +233,7 @@ func findAllFetchedReposWithRef(refText string, n uint64) map[model.SHA1]string 
 		return nil
 	}
 
-	repos := make(map[model.SHA1]string)
+	repos := make(map[model.SHA1]*model.Repository)
 	for rs.Next() {
 		// for each Repository
 		repo, err := rs.Get()
@@ -245,7 +255,7 @@ func findAllFetchedReposWithRef(refText string, n uint64) map[model.SHA1]string 
 			continue
 		}
 
-		repos[masterRef.Init] = repo.ID.String()
+		repos[masterRef.Init] = repo
 	}
 	return repos
 }

--- a/extractor/service.go
+++ b/extractor/service.go
@@ -30,8 +30,14 @@ type Service struct {
 func NewService(n uint64) *Service {
 	//TODO(bzz): parametrize
 	bblfshAddr := "0.0.0.0:9432"
+	grpcMaxMsgSize := 100 * 1024 * 1024
 	log.Info("Connecting to Bblfsh server", "address", bblfshAddr)
-	bblfshConn, err := grpc.Dial(bblfshAddr, grpc.WithTimeout(time.Second*2), grpc.WithInsecure())
+	bblfshConn, err := grpc.Dial(bblfshAddr,
+		grpc.WithTimeout(time.Second*2),
+		grpc.WithInsecure(),
+		grpc.WithDefaultCallOptions(grpc.MaxCallSendMsgSize(grpcMaxMsgSize)),
+		grpc.WithDefaultCallOptions(grpc.MaxCallRecvMsgSize(grpcMaxMsgSize)),
+	)
 	client := protocol.NewProtocolServiceClient(bblfshConn)
 	checkIfError(err)
 

--- a/extractor/service.go
+++ b/extractor/service.go
@@ -22,6 +22,8 @@ import (
 	"gopkg.in/src-d/go-git.v4/storage"
 )
 
+const maxNumOfThousendsOfFilesToProcsee = 10
+
 type Service struct {
 	bblfshClient protocol.ProtocolServiceClient
 	limit        uint64
@@ -53,18 +55,11 @@ func (s *Service) GetRepositoryData(r *Request) (*RepositoryData, error) {
 
 //proteus:generate
 func (s *Service) GetRepositoriesData() ([]*RepositoryData, error) {
-	return s.getRerpoData(s.limit)
+	return s.getRerposData(s.limit)
 }
 
-func (s *Service) getRerpoData(n uint64) ([]*RepositoryData, error) {
-	if n <= 0 {
-		k, err := core.ModelRepositoryStore().Count(model.NewRepositoryQuery().FindByStatus(model.Fetched))
-		if err != nil {
-			log.Error("Could not connect to DB to get the number of 'fetched' repositories", "err", err)
-			return nil, err
-		}
-		n = uint64(k)
-	}
+func (s *Service) getRerposData(n uint64) ([]*RepositoryData, error) {
+	n = allOrN(n)
 	log.Info("Iterating over N repositories in DB", "N", n)
 
 	const master = "refs/heads/master"
@@ -74,7 +69,8 @@ func (s *Service) getRerpoData(n uint64) ([]*RepositoryData, error) {
 	totalFiles := 0
 	for masterRefInit, repoMetadata := range findAllFetchedReposWithRef(master, n) {
 		repo, processedFiles, err := s.processRepository(repoMetadata, master, masterRefInit)
-		if err != nil {
+		if err != nil && repo == nil { //partially processed repos are OK
+			//TODO(bzz): move loggin/error handing here instead of s.processRepository()
 			continue
 		}
 
@@ -115,8 +111,11 @@ func (s *Service) processRepository(repoMetadata *model.Repository, master strin
 	err = tree.Files().ForEach(func(f *object.File) error {
 		i := (skpFiles + sucFiles + errFiles) % 1000
 		batch := (skpFiles + sucFiles + errFiles) / 1000
-		if i == 0 && batch != 0 {
+		if i == 0 && batch != 0 { // CLI progress indicator
 			fmt.Printf("\t%d000 files...\n", batch)
+		}
+		if batch > maxNumOfThousendsOfFilesToProcsee { // only first 10k files
+			return fmt.Errorf("Too many files in a repo. Stopping after 10k")
 		}
 
 		// discard vendoring with enry
@@ -162,7 +161,7 @@ func (s *Service) processRepository(repoMetadata *model.Repository, master strin
 	})
 	if err != nil {
 		log.Error("Failed to iterate files in", "repo", repoID)
-		return nil, sucFiles + errFiles + skpFiles, err
+		return repo, sucFiles + errFiles + skpFiles, err
 	}
 
 	log.Info("Done. All files parsed", "repo", repoID, "success", sucFiles, "fail", errFiles, "skipped", skpFiles)
@@ -266,6 +265,19 @@ func findAllFetchedReposWithRef(refText string, n uint64) map[model.SHA1]*model.
 	return repos
 }
 
+// If N=0, get total number of fetched respoitories from DB. Use N otherwise.
+func allOrN(n uint64) uint64 {
+	const defaultN = 10
+	if n <= 0 {
+		k, err := core.ModelRepositoryStore().Count(model.NewRepositoryQuery().FindByStatus(model.Fetched))
+		if err != nil {
+			log.Error("Cann't connect to DB to get the number of 'fetched' repositories", "err", err)
+			k = defaultN
+		}
+		n = uint64(k)
+	}
+	return n
+}
 func checkIfError(err error) {
 	if err == nil {
 		return

--- a/extractor/service.go
+++ b/extractor/service.go
@@ -23,6 +23,7 @@ import (
 )
 
 const maxNumOfThousendsOfFilesToProcsee = 10
+const	GrpcMaxMsgSize = 100 * 1024 * 1024
 
 type Service struct {
 	bblfshClient protocol.ProtocolServiceClient
@@ -32,13 +33,12 @@ type Service struct {
 func NewService(n uint64) *Service {
 	//TODO(bzz): parametrize
 	bblfshAddr := "0.0.0.0:9432"
-	grpcMaxMsgSize := 100 * 1024 * 1024
 	log.Info("Connecting to Bblfsh server", "address", bblfshAddr)
 	bblfshConn, err := grpc.Dial(bblfshAddr,
 		grpc.WithTimeout(time.Second*2),
 		grpc.WithInsecure(),
-		grpc.WithDefaultCallOptions(grpc.MaxCallSendMsgSize(grpcMaxMsgSize)),
-		grpc.WithDefaultCallOptions(grpc.MaxCallRecvMsgSize(grpcMaxMsgSize)),
+		grpc.WithDefaultCallOptions(grpc.MaxCallSendMsgSize(GrpcMaxMsgSize)),
+		grpc.WithDefaultCallOptions(grpc.MaxCallRecvMsgSize(GrpcMaxMsgSize)),
 	)
 	client := protocol.NewProtocolServiceClient(bblfshConn)
 	checkIfError(err)

--- a/extractor/service.go
+++ b/extractor/service.go
@@ -23,7 +23,7 @@ import (
 )
 
 const maxNumOfThousendsOfFilesToProcsee = 10
-const	GrpcMaxMsgSize = 100 * 1024 * 1024
+const GrpcMaxMsgSize = 100 * 1024 * 1024
 
 type Service struct {
 	bblfshClient protocol.ProtocolServiceClient
@@ -145,7 +145,6 @@ func (s *Service) processRepository(repoMetadata *model.Repository, master strin
 			errFiles++
 			return nil
 		}
-		//log.Debug(fmt.Sprintf("\t%-9s blob %s    %s", fLang, f.Hash, f.Name))
 
 		// Babelfish -> UAST (Python, Java)
 		if strings.EqualFold(fLang, "java") || strings.EqualFold(fLang, "python") {

--- a/normalizer/src/main/scala/tech/sourced/berserker/normalizer/service/ExtractorService.scala
+++ b/normalizer/src/main/scala/tech/sourced/berserker/normalizer/service/ExtractorService.scala
@@ -5,9 +5,13 @@ import io.grpc.ManagedChannelBuilder
 import org.apache.spark.sql.Row
 
 class ExtractorService(host: String, port: Int, isPlainText: Boolean = true) {
+
+  private val maxGrpcMsgSize = 100 * 1042 * 1042
   private val channel = ManagedChannelBuilder
     .forAddress(host, port)
-    .usePlaintext(isPlainText).build()
+    .usePlaintext(isPlainText)
+    .maxInboundMessageSize(maxGrpcMsgSize)
+    .build()
 
   private val stub = ExtractorServiceGrpc.blockingStub(channel)
 

